src/components/Modal/Modal.tsx

```diff
@@ -1,83 +1,110 @@
-import React, { PureComponent } from "react";
+import React, { PureComponent, ReactNode } from "react";
 import { createPortal } from "react-dom";
-import PropTypes from "prop-types";
 import cn from "classnames";
 import Button from "~/components/Button";
 import ModalContext from "./Modal.context";
 import { timeout } from "./constants";
 import "./Modal.css";
 
-let _root;
+let _root: any;
 
-class Modal extends PureComponent {
+type Props = {
+  children: ReactNode;
+  isOpen: boolean;
+  onClose: () => void;
+  fullScreen?: boolean;
+  className: string;
+}
+
+type State = {
+  isMouseClicked: boolean;
+  isAboutToClose: boolean;
+  isClosedGracefully: boolean;
+}
+
+class Modal extends PureComponent<Props, State> {
   static Context = ModalContext;
   static timeout = timeout;
 
-  static propTypes = {
-    children: PropTypes.node,
-    isOpen: PropTypes.bool,
-    onClose: PropTypes.func,
-    fullScreen: PropTypes.bool
-  };
+  state = {
+    isMouseClicked: false,
+    isAboutToClose: false,
 
-  // Initially we assume the state is always that it is closed gracefully.
-  isClosedGracefully = !this.props.isOpen;
+    // Initially we assume the state is always that it is closed gracefully.
+    isClosedGracefully: !this.props.isOpen
+  }
 
-  componentDidUpdate({ isOpen: oldIsOpen }) {
+  onEscape = (e: KeyboardEvent): void => {
+    if (e.key === "Escape") {
+      this.setState({
+        isMouseClicked: true
+      });
+      this.gracefulClose(e);
+    }
+  };
+
+  componentDidUpdate({ isOpen: oldIsOpen }: any): void {
     const newIsOpen = this.props.isOpen;
 
     if (newIsOpen) {
-      this.isClosedGracefully = false;
-      this.isMouseClicked = false;
-
-      this.onEscape = e => {
-        if (e.key === "Escape") {
-          this.isMouseClicked = true;
-          this.gracefulClose();
-        }
-      };
+      this.setState({
+        isClosedGracefully: false,
+        isMouseClicked: false
+      });
 
       window.addEventListener("keyup", this.onEscape);
     }
 
     if (oldIsOpen && !newIsOpen) {
-      this.isAboutToClose = true;
+      this.setState({
+        isAboutToClose: true
+      });
       this.forceUpdate();
 
       setTimeout(() => {
-        this.isAboutToClose = false;
-        this.isClosedGracefully = true;
+        this.setState({
+          isAboutToClose: false,
+          isClosedGracefully: true
+        });
+
         this.forceUpdate();
         window.removeEventListener("keyup", this.onEscape);
       }, timeout);
     }
   }
 
-  componentWillUnmount() {
+  componentWillUnmount(): void {
     window.removeEventListener("keyup", this.onEscape);
   }
 
-  gracefulClose = e => {
-    if (this.isMouseClicked !== true) {
+  gracefulClose = (e: any): void => {
```

Comment:
**suggestion** I think `react` exports `MouseEvent`, maybe we can use that one?

Reply:
Yes, I have to use SyntheticEvent<HTMLDivElement, MouseEvent>